src/base/Quantity.hh

```diff
@@ -0,0 +1,116 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file TaggedUnit.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Macros.hh"
+
+namespace celeritas
+{
+//! Tag class for creating a quantity implicitly from a zero value
+struct ZeroQuantity {};
+//! Get a zero quantity
+inline CELER_CONSTEXPR_FUNCTION ZeroQuantity zero_quantity()
+{
+    return ZeroQuantity{};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * A value tagged with a unit.
+ * \tparam ValueT value type
+ * \tparam UnitT  unit class
+ *
+ * Example usage by physics class, where charge is in units of q_e+, and
+ * mass and momentum are expressed in atomic natural units (where m_e = 1 and c
+ * = 1).
+ * \code
+   using ParticleCharge   = Quantity<real_type, ElChargeUnit>;
+   using ParticleMass     = Quantity<real_type, ElMassUnit>;
+   using ElMomentumUnit   = UnitDivide<ElMassUnit, SpeedOfLight>;
+   using ParticleMomentum = Quantity<real_type, ElMomentumUnit>;
+   \endcode
+ *
+ * A relativistic equation that operates on these quantities can do so without
+ * unnecessary floating point operations involving the speed of light:
+ * \code
+   real_type eval = energy.value(); // Natural units
+   ParticleMomentum momentum{std::sqrt(eval * eval + 2 * mass.value() * eval)};
```

Comment:
`energy` here is kinetic energy, not "kinetic plus rest mass" energy, as is used in the track view. We had some discussion about simplifying "kinetic energy" to "energy" -- it's the natural interpretation for us non-HEP folks, but is it surprising to you?

Reply:
Hm. Maybe a bit. Is it OK to explicitly name it so it's obviously kinetic energy? Otherwise, it's a bit ambiguous. (Though I can adapt.)

Reply:
Yeah, we had this discussion in the past and we settled with making sure that we had clear comments throughout the code as to what it meant. I also thought that it could be confusing at the beginning, but I got used to it pretty quick.